File: lib/trakt/api/TraktAuthApi.kt
package app.moviebase.trakt.api

import app.moviebase.trakt.TraktClientConfig
import app.moviebase.trakt.core.endPoint
import app.moviebase.trakt.model.TraktAccessToken
import app.moviebase.trakt.model.TraktGrantType
import app.moviebase.trakt.model.TraktTokenRefreshRequest
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.HttpRequestBuilder
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.contentType

class TraktAuthApi(
    private val client: HttpClient,
    private val config: TraktClientConfig,
) {
    suspend fun postToken(request: TraktTokenRefreshRequest): TraktAccessToken =
        client
            .post {
                endPointOAuth("token")
                contentType(ContentType.Application.Json)
                setBody(request)
            }.body()

    suspend fun requestAccessToken(
        redirectUri: String,
        code: String,
    ): TraktAccessToken {
        val requestToken =
            TraktTokenRefreshRequest(
                clientId = config.traktApiKey,
                clientSecret = config.clientSecret,
                redirectUri = redirectUri,
                grantType = TraktGrantType.AUTHORIZATION_CODE,
                code = code,
            )
        return postToken(requestToken)
    }

    suspend fun requestRefreshToken(
        redirectUri: String,
        refreshToken: String,
    ): TraktAccessToken {
        require(refreshToken.isNotBlank()) { "refresh token is empty" }

        val requestToken =
            TraktTokenRefreshRequest(
                clientId = config.traktApiKey,
                clientSecret = config.clientSecret,
                redirectUri = redirectUri,
                grantType = TraktGrantType.REFRESH_TOKEN,
                refreshToken = refreshToken,
            )
        return postToken(requestToken)
    }

    private fun HttpRequestBuilder.endPointOAuth(vararg paths: String) {
        endPoint("oauth", *paths)
    }
}

File: lib/services/trakt/trakt_auth_api.dart
import 'package:dio/dio.dart';
import '../../core/config/app_config.dart';
import '../../models/trakt/trakt_auth.dart';

class TraktAuthApi {
  final Dio _dio;
  
  TraktAuthApi(this._dio);
  
  Future<TraktAccessToken> requestAccessToken({
    required String redirectUri,
    required String code,
  }) async {
    final response = await _dio.post(
      '/oauth/token',
      data: {
        'code': code,
        'client_id': AppConfig.traktClientId,
        'client_secret': AppConfig.traktClientSecret,
        'redirect_uri': redirectUri,
        'grant_type': 'authorization_code',
      },
    );
    
    return TraktAccessToken.fromJson(response.data);
  }
  
  Future<TraktAccessToken> refreshAccessToken({
    required String redirectUri,
    required String refreshToken,
  }) async {
    final response = await _dio.post(
      '/oauth/token',
      data: {
        'refresh_token': refreshToken,
        'client_id': AppConfig.traktClientId,
        'client_secret': AppConfig.traktClientSecret,
        'redirect_uri': redirectUri,
        'grant_type': 'refresh_token',
      },
    );
    
    return TraktAccessToken.fromJson(response.data);
  }
  
  Future<void> revokeToken(String token) async {
    await _dio.post(
      '/oauth/revoke',
      data: {
        'token': token,
        'client_id': AppConfig.traktClientId,
        'client_secret': AppConfig.traktClientSecret,
      },
    );
  }
}
